File: src/core/services/client-manager.ts
import { Logger } from 'kv-logger';
import _ from 'lodash';
import { Op } from 'sequelize';
import { Deployments } from '../../models/deployments';
import { DeploymentsVersions } from '../../models/deployments_versions';
import { LogReportDeploy } from '../../models/log_report_deploy';
import { LogReportDownload } from '../../models/log_report_download';
import { Packages } from '../../models/packages';
import { PackagesDiff } from '../../models/packages_diff';
import { PackagesMetrics } from '../../models/packages_metrics';
import { AppError } from '../app-error';
import { config } from '../config';
import { DEPLOYMENT_FAILED, DEPLOYMENT_SUCCEEDED } from '../const';
import { parseVersion, getBlobDownloadUrl } from '../utils/common';
import { redisClient } from '../utils/connections';

const UPDATE_CHECK = 'UPDATE_CHECK';
const CHOSEN_MAN = 'CHOSEN_MAN';
const EXPIRED = 600;

interface UpdateCheckInfo {
    packageId: number;
    downloadURL: string;
    downloadUrl: string;
    description: string;
    isAvailable: boolean;
    isDisabled: boolean;
    isMandatory: boolean;
    appVersion: string;
    targetBinaryRange: string;
    packageHash: string;
    label: string;
    packageSize: number;
    updateAppVersion: boolean;
    shouldRunBinaryVersion: boolean;
    rollout: number;
}

class ClientManager {
    private getUpdateCheckCacheKey(deploymentKey, appVersion, label, packageHash) {
        return [UPDATE_CHECK, deploymentKey, appVersion, label, packageHash].join(':');
    }

    clearUpdateCheckCache(deploymentKey, appVersion, label, packageHash, logger: Logger) {
        logger.info('clear cache Deployments key', {
            key: deploymentKey,
        });
        const redisCacheKey = this.getUpdateCheckCacheKey(
            deploymentKey,
            appVersion,
            label,
            packageHash,
        );
        return redisClient.keys(redisCacheKey).then((data) => {
            if (_.isArray(data)) {
                return Promise.all(
                    data.map((key) => {
                        return redisClient.del(key);
                    }),
                );
            }
            return null;
        });
    }

    updateCheckFromCache(
        deploymentKey: string,
        appVersion: string,
        label: string,
        packageHash: string,
        clientUniqueId: string,
        logger: Logger,
    ) {
        if (!config.common.updateCheckCache) {
            return this.updateCheck(
                deploymentKey,
                appVersion,
                label,
                packageHash,
                clientUniqueId,
                logger,
            );
        }
        const redisCacheKey = this.getUpdateCheckCacheKey(
            deploymentKey,
            appVersion,
            label,
            packageHash,
        );
        return redisClient.get(redisCacheKey).then((data) => {
            if (data) {
                try {
                    logger.debug('updateCheckFromCache read from cache');
                    const obj = JSON.parse(data) as UpdateCheckInfo;
                    return obj;
                } catch (e) {
                    // do nothing
                }
            }
            return this.updateCheck(
                deploymentKey,
                appVersion,
                label,
                packageHash,
                clientUniqueId,
                logger,
            ).then((rs) => {
                try {
                    logger.debug('updateCheckFromCache read from db');
                    const strRs = JSON.stringify(rs);
                    redisClient.setEx(redisCacheKey, EXPIRED, strRs);
                } catch (e) {
                    // do nothing
                }
                return rs;
            });
        });
    }

    private getChosenManCacheKey(packageId, rollout, clientUniqueId) {
        return [CHOSEN_MAN, packageId, rollout, clientUniqueId].join(':');
    }

    private random(rollout) {
        const r = Math.ceil(Math.random() * 10000);
        if (r < rollout * 100) {
            return Promise.resolve(true);
        }
        return Promise.resolve(false);
    }

    chosenMan(packageId, rollout, clientUniqueId: string) {
        if (rollout >= 100) {
            return Promise.resolve(true);
        }
        const rolloutClientUniqueIdCache = _.get(
            config,
            'common.rolloutClientUniqueIdCache',
            false,
        );
        if (rolloutClientUniqueIdCache === false) {
            return this.random(rollout);
        }
        const redisCacheKey = this.getChosenManCacheKey(packageId, rollout, clientUniqueId);
        return redisClient.get(redisCacheKey).then((data) => {
            if (data === '1') {
                return true;
            }
            if (data === '2') {
                return false;
            }
            return this.random(rollout).then((r) => {
                return redisClient
                    .setEx(redisCacheKey, 60 * 60 * 24 * 7, r ? '1' : '2')
                    .then(() => {
                        return r;
                    });
            });
        });
    }

    // eslint-disable-next-line max-lines-per-function
    private async updateCheck(
        deploymentKey: string,
        appVersion: string,
        label: string,
        packageHash: string,
        clientUniqueId: string,
        logger: Logger,
    ): Promise<UpdateCheckInfo | undefined> {
        const startTime = performance.now();

        const rs: UpdateCheckInfo = {
            // ... (Giữ nguyên phần khởi tạo mặc định)
            packageId: 0,
            downloadURL: '',
            downloadUrl: '',
            description: '',
            isAvailable: false,
            isDisabled: true,
            isMandatory: false,
            appVersion,
            targetBinaryRange: '',
            packageHash: '',
            label: '',
            packageSize: 0,
            updateAppVersion: false,
            shouldRunBinaryVersion: false,
            rollout: 100,
        };

        try {
            // 1. Validate & 2. Tìm Deployment & 3. Tìm Version (Giữ nguyên code cũ)
            if (_.isEmpty(deploymentKey) || _.isEmpty(appVersion)) {
                throw new AppError('please input deploymentKey and appVersion');
            }
            const dep = await Deployments.findOne({ where: { deployment_key: deploymentKey } });
            if (!dep) throw new AppError('Not found deployment');

            const version = parseVersion(appVersion);
            const deploymentsVersionsMore = await DeploymentsVersions.findAll({
                where: {
                    deployment_id: dep.id,
                    min_version: { [Op.lte]: version },
                    max_version: { [Op.gt]: version },
                },
            });
            const deploymentsVersions = _.last(_.sortBy(deploymentsVersionsMore, 'created_at'));
            const targetPackageId = _.get(deploymentsVersions, 'current_package_id', 0);

            if (!deploymentsVersions || targetPackageId <= 0) return undefined;

            // 4. Tìm Package đích
            const targetPackage = await Packages.findByPk(targetPackageId);
            if (!targetPackage) return undefined;

            const isSameDeployment =
                targetPackage.deployment_id === deploymentsVersions.deployment_id;
            const isDifferentHash = targetPackage.package_hash !== packageHash;

            if (isSameDeployment && isDifferentHash) {
                // Populate thông tin cơ bản
                rs.packageId = targetPackageId;
                rs.targetBinaryRange = deploymentsVersions.app_version;
                rs.downloadURL = getBlobDownloadUrl(targetPackage.blob_url);
                rs.downloadUrl = rs.downloadURL;
                rs.description = targetPackage.description || '';
                rs.isAvailable = targetPackage.is_disabled !== 1;
                rs.isDisabled = targetPackage.is_disabled === 1;
                rs.isMandatory = targetPackage.is_mandatory === 1;
                rs.appVersion = appVersion;
                rs.packageHash = targetPackage.package_hash;
                rs.label = targetPackage.label;
                rs.packageSize = targetPackage.size;
                rs.rollout = targetPackage.rollout ?? 100;

                // --- [REDIS CACHE LOGIC START] ---
                let finalDescription = rs.description;
                let finalIsMandatory = rs.isMandatory;
                let minId = 0;

                // A. Xác định minId (User đang ở bản nào)
                if (packageHash) {
                    const currentPackage = await Packages.findOne({
                        where: {
                            package_hash: packageHash,
                            deployment_id: targetPackage.deployment_id,
                        },
                    });
                    if (currentPackage) minId = currentPackage.id;
                }

                if (targetPackage.id > minId) {
                    // Tạo Cache Key: Dựa trên ID bản cũ và ID bản mới
                    // Ví dụ: MERGED_INFO:10:20 (Merge từ bản ID 10 lên bản ID 20)
                    const cacheKey = `MERGED_INFO:${packageHash}:${targetPackage.id}`;

                    // B. Thử lấy từ Redis
                    try {
                        const cachedData = await redisClient.get(cacheKey);

                        if (cachedData) {
                            // HIT CACHE: Lấy luôn, không cần query DB nữa
                            logger.debug(`[Cache Hit] ${cacheKey}`);
                            try {
                                const parsed = JSON.parse(cachedData);
                                finalDescription = parsed.description;
                                finalIsMandatory = parsed.isMandatory;
                            } catch (error) {
                                logger.warn(`[Cache] Invalid JSON for ${cacheKey}, deleting...`);
                                redisClient.del(cacheKey);
                            }
                        } else {
                            // MISS CACHE: Query DB và tính toán
                            logger.debug(`[Cache Miss] ${cacheKey} - Querying DB...`);
                            // Kiểm tra mandatory riêng — không giới hạn limit
                            const hasMandatory = await Packages.count({
                                where: {
                                    deployment_id: targetPackage.deployment_id,
                                    id: { [Op.gt]: minId, [Op.lte]: targetPackage.id },
                                    is_disabled: 0,
                                    is_mandatory: 1,
                                },
                            });
                            if (hasMandatory > 0) finalIsMandatory = true;
                            const intermediatePackages = await Packages.findAll({
                                where: {
                                    deployment_id: targetPackage.deployment_id,
                                    id: {
                                        [Op.gt]: minId,
                                        [Op.lte]: targetPackage.id,
                                    },
                                    is_disabled: 0,
                                },
                                order: [['id', 'DESC']],
                                limit: 15,
                            });

                            const messages: string[] = [];
                            intermediatePackages.forEach((pkg) => {
                                if (pkg.description) {
                                    messages.push(`[${pkg.label}]: ${pkg.description}`);
                                }
                            });

                            if (messages.length > 0) {
                                finalDescription = messages.join('\n');
                            }

                            // C. Lưu vào Redis (TTL: 24 giờ = 86400 giây)
                            // Dữ liệu quá khứ không đổi nên có thể cache lâu
                            try {
                                await redisClient.setEx(
                                    cacheKey,
                                    86400,
                                    JSON.stringify({
                                        description: finalDescription,
                                        isMandatory: finalIsMandatory,
                                    }),
                                );
                            } catch (e) {
                                logger.warn(`[Cache] Failed to write ${cacheKey}`, { error: e });
                            }
                        }
                    } catch (cacheError) {
                        logger.warn('[Cache] Redis error, falling back to DB', {
                            error: cacheError,
                        });
                    }
                }

                rs.description = finalDescription;
                rs.isMandatory = finalIsMandatory;

                // 5. Kiểm tra Diff Update
                if (packageHash) {
                    const diffPackage = await PackagesDiff.findOne({
                        where: {
                            package_id: targetPackage.id,
                            diff_against_package_hash: packageHash,
                        },
                    });
                    if (diffPackage) {
                        const diffUrl = getBlobDownloadUrl(diffPackage.diff_blob_url);
                        rs.downloadURL = diffUrl;
                        rs.downloadUrl = diffUrl;
                        rs.packageSize = diffPackage.diff_size;
                    }
                }
            }
            return rs;
        } finally {
            const endTime = performance.now();
            const duration = (endTime - startTime).toFixed(2);
            if (Number(duration) > 200) {
                logger.warn(`[Slow Query] updateCheck took ${duration}ms`, {
                    deploymentKey,
                    appVersion,
                });
            } else {
                logger.debug(`[Perf] updateCheck took ${duration}ms`);
            }
        }
    }

    private getPackagesInfo(deploymentKey, label) {
        if (_.isEmpty(deploymentKey) || _.isEmpty(label)) {
            return Promise.reject(new AppError('please input deploymentKey and label'));
        }
        return Deployments.findOne({ where: { deployment_key: deploymentKey } })
            .then((dep) => {
                if (_.isEmpty(dep)) {
                    throw new AppError('does not found deployment');
                }
                return Packages.findOne({ where: { deployment_id: dep.id, label } });
            })
            .then((packages) => {
                if (_.isEmpty(packages)) {
                    throw new AppError('does not found packages');
                }
                return packages;
            });
    }

    reportStatusDownload(deploymentKey, label, clientUniqueId) {
        return this.getPackagesInfo(deploymentKey, label).then((packages) => {
            return Promise.all([
                PackagesMetrics.findOne({ where: { package_id: packages.id } }).then((metrics) => {
                    if (metrics) {
                        return metrics.increment('downloaded');
                    }
                    return undefined;
                }),
                LogReportDownload.create({
                    package_id: packages.id,
                    client_unique_id: clientUniqueId,
                }),
            ]);
        });
    }

    reportStatusDeploy(deploymentKey: string, label, clientUniqueId: string, others) {
        return this.getPackagesInfo(deploymentKey, label).then((packages) => {
            const statusText = _.get(others, 'status');
            let status = 0;
            if (_.eq(statusText, 'DeploymentSucceeded')) {
                status = DEPLOYMENT_SUCCEEDED;
            } else if (_.eq(statusText, 'DeploymentFailed')) {
                status = DEPLOYMENT_FAILED;
            }
            const packageId = packages.id;
            const previousDeploymentKey = _.get(others, 'previousDeploymentKey');
            const previousLabel = _.get(others, 'previousLabelOrAppVersion');
            if (status > 0) {
                return Promise.all([
                    LogReportDeploy.create({
                        package_id: packageId,
                        client_unique_id: clientUniqueId,
                        previous_label: previousLabel,
                        previous_deployment_key: previousDeploymentKey,
                        status,
                    }),
                    PackagesMetrics.findOne({ where: { package_id: packageId } }).then(
                        (metrics) => {
                            if (_.isEmpty(metrics)) {
                                return undefined;
                            }
                            if (_.eq(status, DEPLOYMENT_SUCCEEDED)) {
                                return metrics.increment(['installed', 'active'], { by: 1 });
                            }
                            return metrics.increment(['installed', 'failed'], { by: 1 });
                        },
                    ),
                ]).then(() => {
                    if (previousDeploymentKey && previousLabel) {
                        return Deployments.findOne({
                            where: { deployment_key: previousDeploymentKey },
                        })
                            .then((dep) => {
                                if (_.isEmpty(dep)) {
                                    return undefined;
                                }
                                return Packages.findOne({
                                    where: { deployment_id: dep.id, label: previousLabel },
                                }).then((p) => {
                                    if (_.isEmpty(p)) {
                                        return undefined;
                                    }
                                    return PackagesMetrics.findOne({
                                        where: { package_id: p.id },
                                    });
                                });
                            })
                            .then((metrics) => {
                                if (metrics) {
                                    return metrics.decrement('active');
                                }
                                return undefined;
                            });
                    }
                    return undefined;
                });
            }
            return undefined;
        });
    }
}

export const clientManager = new ClientManager();

File: rust/src/services/client_manager.rs
use sqlx::SqlitePool;
use bb8_redis::{bb8, RedisConnectionManager, redis::AsyncCommands};
use serde::{Deserialize, Serialize};
use rand::Rng;

use crate::core::app_error::AppError;
use crate::models::deployments::Deployment;
use crate::models::deployments_versions::DeploymentVersion;
use crate::models::packages::Packages;
use crate::models::packages_diff::PackagesDiff;
use crate::utils::common::{parse_version, get_blob_download_url};

#[derive(Serialize, Deserialize)]
pub struct UpdateCheckInfo {
    pub package_id: i64,
    pub download_url: String,
    pub description: String,
    pub is_available: bool,
    pub is_disabled: bool,
    pub is_mandatory: bool,
    pub app_version: String,
    pub target_binary_range: String,
    pub package_hash: String,
    pub label: String,
    pub package_size: i64,
    pub rollout: i64,
}

pub struct ClientManager;

impl ClientManager {
    async fn compute_merged_info(
        pool: &SqlitePool,
        deployment_id: i64,
        deployment_version_id: i64,
        min_id: i64,
        target_id: i64,
    ) -> Result<(String, bool), AppError> {
        let has_mandatory: i64 = sqlx::query_scalar(
            "SELECT count(*) FROM packages WHERE deployment_id = ? AND deployment_version_id = ? AND id > ? AND id <= ? AND is_disabled = 0 AND is_mandatory = 1"
        )
        .bind(deployment_id)
        .bind(deployment_version_id)
        .bind(min_id)
        .bind(target_id)
        .fetch_one(pool)
        .await?;

        let intermediate_packages = sqlx::query_as::<_, Packages>(
            "SELECT * FROM packages WHERE deployment_id = ? AND deployment_version_id = ? AND id > ? AND id <= ? AND is_disabled = 0 ORDER BY id DESC LIMIT 15"
        )
        .bind(deployment_id)
        .bind(deployment_version_id)
        .bind(min_id)
        .bind(target_id)
        .fetch_all(pool)
        .await?;

        let mut messages = Vec::new();
        for pkg in intermediate_packages {
            if !pkg.description.is_empty() {
                let pkg_label = if pkg.label.is_empty() { "unknown" } else { &pkg.label };
                messages.push(format!("[{}]: {}", pkg_label, pkg.description));
            }
        }

        let description = if messages.is_empty() {
            "".to_string()
        } else {
            messages.join("\n")
        };

        Ok((description, has_mandatory > 0))
    }

    pub async fn clear_update_check_cache(
        redis: &bb8_redis::bb8::Pool<bb8_redis::RedisConnectionManager>,
        deployment_key: &str,
    ) {
        let cache_key = format!("UPDATE_CHECK:{}:*", deployment_key);
        if let Ok(mut conn) = redis.get().await {
            use bb8_redis::redis::AsyncCommands;
            if let Ok(keys) = conn.keys::<_, Vec<String>>(&cache_key).await {
                for key in keys {
                    let _: () = conn.del(key).await.unwrap_or(());
                }
            }
        }
    }

    pub async fn update_check(
        state: &crate::core::db::AppState,
        deployment_key: &str,
        app_version: &str,
        label: &str,
        package_hash: &str,
        _client_unique_id: &str,
    ) -> Result<Option<UpdateCheckInfo>, AppError> {
        let pool = &state.db.pool;
        let redis = &state.db.redis;
        let use_cache = state.config.update_check_cache;
        if deployment_key.is_empty() || app_version.is_empty() {
            return Err(AppError::new("please input deploymentKey and appVersion"));
        }

        let cache_key = format!("UPDATE_CHECK:{}:{}:{}:{}", deployment_key, app_version, label, package_hash);
        let mut conn = redis.get().await.map_err(|e| AppError::new(&format!("Redis error: {}", e)))?;
        
        if use_cache {
            let cached_data: Option<String> = bb8_redis::redis::AsyncCommands::get(&mut *conn, &cache_key).await.unwrap_or(None);
            if let Some(data) = cached_data {
                if data == "null" {
                    return Ok(None);
                }
                if let Ok(info) = serde_json::from_str::<UpdateCheckInfo>(&data) {
                    return Ok(Some(info));
                }
            }
        }

        let dep = sqlx::query_as::<_, Deployment>(
            "SELECT * FROM deployments WHERE deployment_key = ?"
        )
        .bind(deployment_key)
        .fetch_optional(pool)
        .await?
        .ok_or_else(|| AppError::new("Not found deployment"))?;

        let version = parse_version(app_version);

        let deployments_versions = sqlx::query_as::<_, DeploymentVersion>(
            "SELECT * FROM deployments_versions WHERE deployment_id = ? AND min_version <= ? AND max_version > ? ORDER BY created_at DESC LIMIT 1"
        )
        .bind(dep.id)
        .bind(&version)
        .bind(&version)
        .fetch_optional(pool)
        .await?;

        let deployments_version = match deployments_versions {
            Some(dv) => dv,
            None => return Ok(None),
        };

        let target_package_id = deployments_version.current_package_id;
        if target_package_id <= 0 {
            return Ok(None);
        }

        let target_package = sqlx::query_as::<_, Packages>(
            "SELECT * FROM packages WHERE id = ?"
        )
        .bind(target_package_id)
        .fetch_optional(pool)
        .await?;

        let target_package = match target_package {
            Some(p) => p,
            None => return Ok(None),
        };

        let is_same_deployment = target_package.deployment_id == deployments_version.deployment_id;
        let is_different_hash = target_package.package_hash != package_hash;

        if is_same_deployment && is_different_hash {
            let mut rs = UpdateCheckInfo {
                package_id: target_package.id,
                target_binary_range: deployments_version.app_version.clone(),
                download_url: get_blob_download_url(&target_package.blob_url),
                description: target_package.description.clone(),
                is_available: target_package.is_disabled == 0,
                is_disabled: target_package.is_disabled == 1,
                is_mandatory: target_package.is_mandatory == 1,
                app_version: app_version.to_string(),
                package_hash: target_package.package_hash.clone(),
                label: target_package.label.clone(),
                package_size: target_package.size as i64,
                rollout: target_package.rollout as i64,
            };

            let mut final_description = rs.description.clone();
            let mut final_is_mandatory = rs.is_mandatory;
            let mut min_id = 0;

            if !package_hash.is_empty() {
                let current_package = sqlx::query_as::<_, Packages>(
                    "SELECT * FROM packages WHERE package_hash = ? AND deployment_id = ?"
                )
                .bind(package_hash)
                .bind(target_package.deployment_id)
                .fetch_optional(pool)
                .await?;
                
                if let Some(cp) = current_package {
                    min_id = cp.id;
                }
            } else {
                min_id = target_package.id - 1;
            }

            if target_package.id > min_id {
                let merged_cache_key = format!("MERGED_INFO:{}:{}:{}", deployment_key, package_hash, target_package.id);
                let merged_data: Option<String> = conn.get(&merged_cache_key).await.unwrap_or(None);
                
                let mut hit = false;
                if let Some(data) = merged_data {
                    if let Ok(parsed) = serde_json::from_str::<serde_json::Value>(&data) {
                        if let (Some(desc), Some(mand)) = (parsed.get("description").and_then(|v| v.as_str()), parsed.get("isMandatory").and_then(|v| v.as_bool())) {
                            final_description = desc.to_string();
                            final_is_mandatory = mand;
                            hit = true;
                        }
                    }
                }

                if !hit {
                    let (desc, mand) = Self::compute_merged_info(pool, target_package.deployment_id, deployments_version.id, min_id, target_package.id).await?;
                    final_description = desc.clone();
                    final_is_mandatory = mand;

                    let cache_val = serde_json::json!({
                        "description": desc,
                        "isMandatory": mand
                    });
                    let _: () = conn.set_ex(&merged_cache_key, cache_val.to_string(), 86400).await.unwrap_or(());
                }
            }

            rs.description = final_description;
            rs.is_mandatory = final_is_mandatory;

            // diff check
            if !package_hash.is_empty() {
                let diff_package = sqlx::query_as::<_, PackagesDiff>(
                    "SELECT * FROM packages_diff WHERE package_id = ? AND diff_against_package_hash = ?"
                )
                .bind(target_package.id)
                .bind(package_hash)
                .fetch_optional(pool)
                .await?;

                if let Some(diff) = diff_package {
                    rs.download_url = get_blob_download_url(&diff.diff_blob_url);
                    rs.package_size = diff.diff_size as i64;
                }
            }

            if let Ok(json) = serde_json::to_string(&rs) {
                if use_cache {
                    let _: () = bb8_redis::redis::AsyncCommands::set_ex(&mut *conn, &cache_key, json, 600).await.unwrap_or(());
                }
            }

            return Ok(Some(rs));
        }

        if use_cache {
            let _: () = bb8_redis::redis::AsyncCommands::set_ex(&mut *conn, &cache_key, "null", 600).await.unwrap_or(());
        }
        Ok(None)
    }

    pub async fn report_status_download(
        pool: &SqlitePool,
        deployment_key: &str,
        label: &str,
        client_unique_id: &str,
    ) -> Result<(), AppError> {
        let dep = sqlx::query_as::<_, Deployment>(
            "SELECT * FROM deployments WHERE deployment_key = ?"
        )
        .bind(deployment_key)
        .fetch_optional(pool)
        .await?
        .ok_or_else(|| AppError::new("does not found deployment"))?;

        let package = sqlx::query_as::<_, Packages>(
            "SELECT * FROM packages WHERE deployment_id = ? AND label = ?"
        )
        .bind(dep.id)
        .bind(label)
        .fetch_optional(pool)
        .await?
        .ok_or_else(|| AppError::new("does not found packages"))?;

        let mut tx = pool.begin().await?;

        sqlx::query(
            "UPDATE packages_metrics SET downloaded = downloaded + 1 WHERE package_id = ?"
        )
        .bind(package.id)
        .execute(&mut *tx)
        .await?;

        sqlx::query(
            "INSERT INTO log_report_download (package_id, client_unique_id, created_at) VALUES (?, ?, CURRENT_TIMESTAMP)"
        )
        .bind(package.id)
        .bind(client_unique_id)
        .execute(&mut *tx)
        .await?;

        tx.commit().await?;

        Ok(())
    }

    pub async fn report_status_deploy(
        pool: &SqlitePool,
        deployment_key: &str,
        label: &str,
        client_unique_id: &str,
        status_opt: Option<&str>,
        previous_deployment_key: Option<&str>,
        previous_label_or_app_version: Option<&str>,
    ) -> Result<(), AppError> {
        let dep = sqlx::query_as::<_, Deployment>(
            "SELECT * FROM deployments WHERE deployment_key = ?"
        )
        .bind(deployment_key)
        .fetch_optional(pool)
        .await?
        .ok_or_else(|| AppError::new("does not found deployment"))?;

        let package = sqlx::query_as::<_, Packages>(
            "SELECT * FROM packages WHERE deployment_id = ? AND label = ?"
        )
        .bind(dep.id)
        .bind(label)
        .fetch_optional(pool)
        .await?
        .ok_or_else(|| AppError::new("does not found packages"))?;

        let status = match status_opt {
            Some("DeploymentSucceeded") => 1,
            Some("DeploymentFailed") => 2,
            _ => 0,
        };

        if status > 0 {
            let mut tx = pool.begin().await?;

            sqlx::query(
                "INSERT INTO log_report_deploy (package_id, client_unique_id, previous_label, previous_deployment_key, status, created_at) VALUES (?, ?, ?, ?, ?, CURRENT_TIMESTAMP)"
            )
            .bind(package.id)
            .bind(client_unique_id)
            .bind(previous_label_or_app_version)
            .bind(previous_deployment_key)
            .bind(status)
            .execute(&mut *tx)
            .await?;

            if status == 1 {
                sqlx::query(
                    "UPDATE packages_metrics SET installed = installed + 1, active = active + 1 WHERE package_id = ?"
                )
                .bind(package.id)
                .execute(&mut *tx)
                .await?;
            } else {
                sqlx::query(
                    "UPDATE packages_metrics SET installed = installed + 1, failed = failed + 1 WHERE package_id = ?"
                )
                .bind(package.id)
                .execute(&mut *tx)
                .await?;
            }

            if let (Some(prev_dep_key), Some(prev_label)) = (previous_deployment_key, previous_label_or_app_version) {
                let prev_dep = sqlx::query_as::<_, Deployment>(
                    "SELECT * FROM deployments WHERE deployment_key = ?"
                )
                .bind(prev_dep_key)
                .fetch_optional(&mut *tx)
                .await?;

                if let Some(p_dep) = prev_dep {
                    let prev_pkg = sqlx::query_as::<_, Packages>(
                        "SELECT * FROM packages WHERE deployment_id = ? AND label = ?"
                    )
                    .bind(p_dep.id)
                    .bind(prev_label)
                    .fetch_optional(&mut *tx)
                    .await?;

                    if let Some(p_pkg) = prev_pkg {
                        sqlx::query(
                            "UPDATE packages_metrics SET active = active - 1 WHERE package_id = ?"
                        )
                        .bind(p_pkg.id)
                        .execute(&mut *tx)
                        .await?;
                    }
                }
            }

            tx.commit().await?;
        }

        Ok(())
    }

    pub async fn chosen_man(
        _pool: &SqlitePool,
        redis: &bb8::Pool<RedisConnectionManager>,
        package_id: i64,
        rollout: Option<i64>,
        client_unique_id: &str,
    ) -> Result<bool, AppError> {
        let rollout = rollout.unwrap_or(100);
        if rollout >= 100 {
            return Ok(true);
        }

        let cache_key = format!("CHOSEN_MAN:{}:{}:{}", package_id, rollout, client_unique_id);
        let mut conn = redis.get().await.map_err(|e| AppError::new(&format!("Redis error: {}", e)))?;
        
        let data: Option<String> = conn.get(&cache_key).await.unwrap_or(None);
        if let Some(val) = data {
            if val == "1" {
                return Ok(true);
            }
            if val == "2" {
                return Ok(false);
            }
        }

        let chosen = {
            let mut rng = rand::thread_rng();
            rng.gen_range(0..10000) < rollout * 100
        };

        let val = if chosen { "1" } else { "2" };
        let _: () = conn.set_ex(&cache_key, val, 60 * 60 * 24 * 7).await.unwrap_or(());

        Ok(chosen)
    }
}
